Add Feature - Add NPKit Support in RCCL

yzygitzh opened the pull request:
This commit enables RCCL to dump NPKit-compatible trace file. The main changes include:

1) NPKit interface and core functionalities, related changed files include:
src/include/npkit/npkit.h
src/include/npkit/npkit_event.h
src/include/npkit/npkit_struct.h
src/misc/npkit.cc

2) NPKit initialization and destroy, related changed files include:
src/include/devcomm.h
src/include/proxy.h
src/init.cc

3) NPKit tracing logic implementation in COLL operations and networking proxy, related changed files include:
src/collectives/device/all_reduce.h
src/collectives/device/prims_ll.h
src/collectives/device/prims_simple.h
src/collectives/device/sendrecv.h
src/transport/net.cc

4) Build script change:
CMakeLists.txt

wenkaidu commented:
Can you add a NPKit how to document? Also since the commit doesn't include the "sample" folder, we need "sample" folder from NPKit github to process the results which are diff files at this time.

yzygitzh commented:
> Can you add a NPKit how to document? Also since the commit doesn't include the "sample" folder, we need "sample" folder from NPKit github to process the results which are diff files at this time.

Sure, let me do it.

yzygitzh commented:
> Can you add a NPKit how to document? Also since the commit doesn't include the "sample" folder, we need "sample" folder from NPKit github to process the results which are diff files at this time.

Added NPKit instructions into README. Also added a pointer to `rccl_samples` in NPKit repo, it's a fully automated example with post-process script in it.
